src/main.cpp

```diff
@@ -43,32 +53,91 @@ int main(int argc, char *argv[]) {
+                    ml::util::loadFromFile<aiTools::State::FEATURE_VEC_LEN, 200, 200, 1>(*pretrainedNet)));
+    } else {
+        mlps = std::make_unique<std::pair<ml::Mlp<aiTools::State::FEATURE_VEC_LEN, 200, 200, 1>,
+            ml::Mlp<aiTools::State::FEATURE_VEC_LEN, 200, 200, 1>>>(std::make_pair<ml::Mlp<aiTools::State::FEATURE_VEC_LEN, 200, 200, 1>,
             ml::Mlp<aiTools::State::FEATURE_VEC_LEN, 200, 200, 1>>({ml::functions::relu, ml::functions::relu, ml::functions::identity},
-                                                                   {ml::functions::relu, ml::functions::relu, ml::functions::identity});
+                                                                   {ml::functions::relu, ml::functions::relu, ml::functions::identity}));
+    }
+
+    if(argc < 8){
+        log.info("No directory for experience replay specified");
+    }
+
+    if(expDirList.has_value() && expDirList->empty()){
+        log.warn("No experiences found in " + std::string(argv[7]));
+    } else {
+        log.warn("Found " + std::to_string(expDirList->size()) + " directories in " + std::string(argv[7]));
+        expReplayEnabled = true;
+        dirListIt = expDirList->begin();
+        expDirIt = std::filesystem::directory_iterator(*dirListIt);
+    }
+
 
     for (auto epoch = 0; epoch < std::numeric_limits<int>::max(); ++epoch) {
-        communication::Communicator communicator{matchConfig, leftTeamConfig, rightTeamConfig, log, learningRate,
-                                                 discountRate, mlps};
+        std::unique_ptr<Communicator> communicator;
+        if(expReplayEnabled && epoch % *expEpochs == 0){
+            if(*expDirIt == std::filesystem::end(*expDirIt)) {
+                if(++dirListIt == expDirList->end()){
+                    log.warn("--- No experience left, resetting ---");
+                    std::exit(0);
+                    dirListIt = expDirList->begin();
+                }
 
-        mlps = communicator.mlps;
+                log.warn("--- Current experience exhausted, fetching next directory ---");
+                expDirIt.emplace(std::filesystem::directory_iterator(*dirListIt));
+            }
+
+            log.warn("--- Experience replay epoch ---");
+            auto state = readFromFileToJson<aiTools::State>(expDirIt.value()->path().string());
+            communicator = std::make_unique<Communicator>(matchConfig, state, log, learningRate, discountRate, *mlps, "---");
+            ++(*expDirIt);
+
+        } else {
+            communicator = std::make_unique<Communicator>(matchConfig, leftTeamConfig, rightTeamConfig, log, learningRate,
+                                                 discountRate, *mlps, "---");
+        }
 
+        mlps = std::make_unique<decltype(communicator->mlps)>(communicator->mlps);
         log.warn("Epoch finished: " + std::to_string(epoch));
 
-        if (epoch % 100000 == 0) {
+        if (epoch % 10000 == 0) {
```

Comment:
kann man auch auslagern statt hart in den code aber ist an der Stelle jetzt auch nicht dramatisch